### PR TITLE
Allow the observer designation of 'self' for a HeliographicCarrington coordinate

### LINKED ISSUE #4279: Enable a HeliographicCarrington frame to be defined using only the Sun-observer distance
Our `HeliographicCarrington` frame is now observer-based (#3782), and the API requires a complete coordinate for the observer.  This turns out to be awkward when wanting to create a `HeliographicCarrington` coordinate representing an observer location, because the observer coordinate is seemingly needed to define the frame for the observer coordinate.  Mathematically, the `HeliographicCarrington`->`HeliographicStonyhurst` transformation needs only the Sun-observer distance, and modifies only the longitude component, so the circular loop is solely a consequence of the implementation.

There should be a way to define a `HeliographicCarrington` frame using only the Sun-observer distance, as opposed to a complete coordinate for the observer.  Either a new frame attribute can be added or our existing `ObserverCoordinateAttribute` can be extended.

### PR DESCRIPTION
**Pull request by ayshih**
Closes #4279

Some data sets specify the observer location in Carrington coordinates.  Following the change in #3782 where our `HeliographicCarrington` frame requires the observer location to be specified, this resulted in an annoying circular loop in creating such coordinates (see #4273 and #4279).  This PR creates a special observer designation (`observer='self'`) that can be used with `HeliographicCarrington`.  This approach also enables removing hack-ish code in `Map`.

```python
>>> from astropy.coordinates import SkyCoord
>>> import astropy.units as u

>>> sc = SkyCoord(0*u.deg, 0*u.deg, 1*u.AU,
...               frame='heliographic_carrington', obstime='2001-01-01', observer='self')
>>> sc
<SkyCoord (HeliographicCarrington: obstime=2001-01-01T00:00:00.000, observer=self): (lon, lat, radius) in (deg, deg, AU)
    (0., 0., 1.)>

>>> sc.heliographic_stonyhurst
<SkyCoord (HeliographicStonyhurst: obstime=2001-01-01T00:00:00.000): (lon, lat, radius) in (deg, deg, AU)
    (142.71381173, 0., 1.)>
```

### CODE REVIEW

**Comment by Cadair** [comment]
Looks like a good solution.

**Comment by ayshih** [comment]
@dstansby Can you test whether this resolves your use case?

**Comment by dstansby** [comment]
I'm hitting https://github.com/sunpy/sunpy/blob/834eb0269f6a3acda080ecdb8131e7d60397bc1f/sunpy/coordinates/transformations.py#L202-L204
when trying to transform 
```python
<SkyCoord (HeliographicCarrington: obstime=[datetime.datetime(2020, 11, 20, 12, 0)], observer=self): (lon, lat, radius) in (deg, deg, m)
    [(72.66343313, 3.64720498, 1.732293e+09)]>
```
to 
```python
<HeliographicCarrington Frame (obstime=2020-11-20T12:00:00.000, observer=<HeliographicStonyhurst Coordinate (obstime=2020-11-20T12:00:00.000): (lon, lat, radius) in (deg, deg, m)
    (0., 2.34419858, 1.47847545e+11)>)>
```
so it's not working for my use case. Perhaps `'self'` needs to resolve more eagerly into an actual coordinate, or an additional check needs to go in `_observers_are_equal`?

**Comment by ayshih** [comment]
Okay, the HCC loopback transformation is now fixed.  Can you try again?

**Comment by dstansby** [comment]
Thanks, tried and it's working a treat 👍

**Review by Cadair — COMMENTED** [review]
Looks good to me, although I like the idea of disallowing the specification of the observer as self on anything other than a HGC frame. Though I could well be missing something as to why that wouldn't work.

File: sunpy/coordinates/transformations.py [hunk 1]
```diff
@@ -217,9 +221,13 @@ def _check_observer_defined(frame):
...
-        raise ConvertError("This transformation cannot be performed because the "
-                           f"{frame.__class__.__name__} frame needs a specified obstime "
-                           f"to fully resolve observer='{frame.observer}'.")
+        if frame.observer != "self":
+            raise ConvertError("This transformation cannot be performed because the "
+                               f"{frame.__class__.__name__} frame needs a specified obstime "
+                               f"to fully resolve observer='{frame.observer}'.")
+        elif not isinstance(frame, HeliographicCarrington):
+            raise ConvertError(f"The {frame.__class__.__name__} frame has observer='self' "
+                               "but this is valid for only HeliographicCarrington frames.")
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/transformations.py:230]: I actually don't think it's possible the way you envision.  I don't think the frame attribute is aware of the frame class that it belongs to when it is being set (which is not through a `__set__()` method).  You could error on the `__get__()` method, which does know the frame class, but that's too late for what you want.

Besides, technically speaking, there's nothing *wrong* with `observer='self'` for any observer-based frame, and arguably it's more complete a definition than `observer=None`, which we support.  We just don't know how to transform an observer-based frame with `observer='self'` for anything other than HGC, and so this PR does raise an error once someone tries such a transformation.

[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/transformations.py:230]: A correction: `observer='self'` is only sometimes techinally sensible for the non-HGC frames:
* For `Heliocentric`, it is sensible only along the Sun-observer line (the Z-axis)
* For `Helioprojective`, it is sensible only at the origin

[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/coordinates/transformations.py:230]: Descriptors are weird. I guess we could do this by catching setting of the hidden `frame._observer` attribute? Probably not worth the faff.

**Review by Cadair — APPROVED** [review]

**Pull request merged**